Java example for "Grid with sortable columns" throws an exception when sort order is cleared

@fluorumlabs commented on [Fri Aug 03 2018](https://github.com/vaadin/flow/issues/4482)

Steps to reproduce:
1. Open https://vaadin.com/components/vaadin-grid/java-examples/sorting
2. Click any column three times
3. On the third click exception is thrown and page is reloaded

```
Caused by: java.lang.ClassCastException: elemental.json.impl.JreJsonNull cannot be cast to elemental.json.JsonString
	at elemental.json.impl.JreJsonObject.getString(JreJsonObject.java:113)
	at com.vaadin.flow.component.grid.Grid.sortersChanged(Grid.java:2013)
```

---

@Legioth commented on [Tue Aug 07 2018](https://github.com/vaadin/flow/issues/4482#issuecomment-410980446)

Already fixed in Grid 2.0 through https://github.com/vaadin/vaadin-grid-flow/pull/249. Should maybe be backported to the 1.0 series of Grid?

---

@denis-anisimov commented on [Sat Aug 18 2018](https://github.com/vaadin/flow/issues/4482#issuecomment-414049436)

I cannot reproduce it with the latest version.

I've added an IT test for this case and one more check in the code to avoid exception if `direction` value is `null`. May be this change is extra but anyway there is an IT test.

--- a/src/main/java/com/vaadin/flow/component/grid/Grid.java
+++ b/src/main/java/com/vaadin/flow/component/grid/Grid.java
@@ -2210,16 +2210,18 @@ public class Grid<T> extends Component implements HasDataProvider<T>, HasStyle,
                 throw new IllegalArgumentException(
                         "Received a sorters changed call from the client for a non-existent column");
             }
-            switch (sorter.getString("direction")) {
-            case "asc":
-                sortOrderBuilder.thenAsc(column);
-                break;
-            case "desc":
-                sortOrderBuilder.thenDesc(column);
-                break;
-            default:
-                throw new IllegalArgumentException(
-                        "Received a sorters changed call from the client containing an invalid sorting direction");
+            if (sorter.hasKey("direction")) {
+                switch (sorter.getString("direction")) {
+                case "asc":
+                    sortOrderBuilder.thenAsc(column);
+                    break;
+                case "desc":
+                    sortOrderBuilder.thenDesc(column);
+                    break;
+                default:
+                    throw new IllegalArgumentException(
+                            "Received a sorters changed call from the client containing an invalid sorting direction");
+                }
             }
         }
         setSortOrder(sortOrderBuilder.build(), true);

--- a/src/test/java/com/vaadin/flow/component/grid/it/GridViewIT.java
+++ b/src/test/java/com/vaadin/flow/component/grid/it/GridViewIT.java
@@ -433,6 +433,21 @@ public class GridViewIT extends TabbedComponentDemoTest {
                 QuerySortOrder.asc("age").thenAsc("name").build(), true);
     }
 
+    @Test
+    public void gridWithSorting_switchColumnSorting() {
+        openTabAndCheckForErrors("sorting");
+        GridElement grid = $(GridElement.class).id("grid-sortable-columns");
+        scrollToElement(grid);
+
+        getCellContent(grid.getHeaderCell(0)).click();
+        getCellContent(grid.getHeaderCell(0)).click();
+        getCellContent(grid.getHeaderCell(0)).click();
+
+        Assert.assertEquals(
+                "Current sort order: . Sort originates from the client: true.",
+                findElement(By.id("grid-sortable-columns-message")).getText());
+    }
+
     private void assertSortMessageEquals(List<QuerySortOrder> querySortOrders,
             boolean fromClient) {
         String sortOrdersString = querySortOrders.stream()

